### PR TITLE
Fix regloss logging

### PR DESCRIPTION
**Pull request by AkshitaB**
Fixes https://github.com/allenai/allennlp/issues/4436 

reg_loss is only displayed if the model being trained is configured to have a regularization penalty.

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -409,6 +409,11 @@ def __init__(
         self._tensorboard.get_batch_num_total = lambda: self._batch_num_total
         self._tensorboard.enable_activation_logging(self.model)
 
+        # Only display reg_loss if the model's configuration has regularization.
+        self._show_metrics = ["loss", "reg_loss"]
+        if self.model.get_regularization_penalty() == 0.0:
```
Comment (dirkgr) [inline allennlp/training/trainer.py:414]: It's likely that we have to detect whether a regularization loss is configured by looking whether the number is `0`. But it would be better if we make that decision based on whether it's configured, not based on whether the result is zero. That way we can show it, for example, when it's misconfigured.

Comment (AkshitaB) [reply allennlp/training/trainer.py:414]: I believe the `get_regularization_penalty()` function returns a 0 if the model is configured to not have regularization. This is what the docstring for that function says: `Returns 0 if the model was not configured to use regularization.`

**Comment by matt-gardner** [comment]
I agree that what @dirkgr suggests sounds like a cleaner change.

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -574,19 +578,20 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
             for batch in batch_group:
                 batch_outputs = self.batch_outputs(batch, for_training=True)
                 batch_group_outputs.append(batch_outputs)
-                loss = batch_outputs["loss"]
-                reg_loss = batch_outputs["reg_loss"]
+                loss = batch_outputs.get("loss")
```
Comment (dirkgr) [inline allennlp/training/trainer.py:581]: I think `"loss"` is always there.

File: allennlp/models/model.py [hunk 3]
```diff
@@ -78,15 +78,18 @@ def __init__(self, vocab: Vocabulary, regularizer: RegularizerApplicator = None)
...
+        Returns None if the model was not configured to use regularization.
         """
         if self._regularizer is None:
-            return 0.0
+            regularization_penalty = None
         else:
-            return self._regularizer(self)
+            regularization_penalty = self._regularizer(self)
+            if isinstance(regularization_penalty, float):
+                regularization_penalty = torch.tensor(regularization_penalty)
```
Comment (dirkgr) [inline allennlp/models/model.py:91]: If we get a `float` from `self._regularizer`, how does that work with gradients? I don't see how regularization can work if we lose track of the computation graph here. `float` doesn't have any information about the computation graph.

I guess it works if the `float` is 0. Is that what this case is capturing?

File: allennlp/training/trainer.py (hunk 2 above)
Comment (AkshitaB) [reply allennlp/training/trainer.py:581]: Yes, it was more for uniformity's sake.

File: allennlp/models/model.py (hunk 3 above)
Comment (AkshitaB) [reply allennlp/models/model.py:91]: That's what I also thought. But this test case: `TestElmoTokenEmbedder.test_tagger_with_elmo_token_embedder_can_train_save_and_load` seems to require explicit conversion to tensor. The value is actually 0.0 .

Comment (dirkgr) [reply allennlp/models/model.py:91]: I see. This happens in the applicator when none of the regexes match. That's probably not a great situation, since it might mean the config doesn't do what you think it does. But let's not invent a whole new problem area to solve in this PR.

In this case, let's throw an exception when the number isn't `0.0` and call it done?
